### PR TITLE
feature/action deploy

### PR DESCRIPTION
**Pull request by duncanhealy**
## Description

add action to publish docker image to docker hub

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Related Issue

#3

## Motivation and Context

automates creation of docker hub image

## Checklist:

- [ ] I have updated the documentation (if required).
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] I added a picture of a cute animal cause it's fun
^ we need a bot to add a cute animal when prs land

### CODE REVIEW

**Comment by duncanhealy** [comment]
publish to github container registry to add too?
 
- [ ] relevant details {{ secrets.DOCKERHUB_TOKEN }} {{ secrets.DOCKERHUB_USERNAME }} need to be added to the repo  
match the owner/user there (I put owasp/cwe-tool) not sure what dockerhub user you are using

- [ ] I will update pr to add multiple tags on it 

also could always add a certain security scanner to check the docker image before it is pushed, wonder where we could find one of those

**Comment by lirantal** [comment]
Cool, good stuff here!
Should we land #3 first, and then continue with this one?

**Comment by duncanhealy** [comment]
ok pushes to both container registry once the DOCKER environment is set up 
That should fix the failing step above 
The name on the github package registry (GPR) is OWASP/cwe-tool/cwe-tool
I've added doc to the Readme on how to pull down the image from the GPR 
the repository should be marked public (not private)

**Comment by lirantal** [comment]
@duncanhealy a lot of versioning and github actions kung-fu there :-)
So what does this essentially do? Seems like:
1. Build a docker image
2. Pushes it to Docker Hub
3. Pushers it to GitHub Package Registry (which is essentially a docker container registry).

Is that the case?

Also, can we just name it `OWASP/cwe-tool` for GitHub Registry instead of the verbose name `OWASP/cwe-tool/cwe-tool` which might look as if someone had a typo?

And finally, I think the README can change so that:
1. First it instructs users to either use the image from docker hub or from github registry
2. Then it explains how to build and run locally
Sounds good?

**Comment by duncanhealy** [comment]
> 1. Build a docker image
> 2. Pushes it to Docker Hub
> 3. Pushers it to GitHub Package Registry (which is essentially a docker container registry).
> 
> Is that the case?

:+1: yup

> Also, can we just name it `OWASP/cwe-tool` for GitHub Registry instead of the verbose name `OWASP/cwe-tool/cwe-tool` which might look as if someone had a typo?

we would have to switch to container registry for that 
https://docs.github.com/en/free-pro-team@latest/packages/getting-started-with-github-container-registry/migrating-to-github-container-registry-for-docker-images
default for package registry is form ```PACKAGE-TYPE.pkg.github.com/OWNER/REPOSITORY/IMAGE-NAME```

you would need to set another secret on the repo a PAT - https://docs.github.com/en/free-pro-team@latest/github/authenticating-to-github/creating-a-personal-access-token

> And finally, I think the README can change so that:
> 
> 1. First it instructs users to either use the image from docker hub or from github registry
> 2. Then it explains how to build and run locally
>    Sounds good?

sounds good - I can update the README

**Comment by lirantal** [comment]
@duncanhealy makes sense to update the PR then to push the docker image to GitHub's Registry?

Not sure why did we add this:
`        if: ${{ secrets.DOCKERHUB_USERNAME=="owasp" }}`
I'd probably need to use my own private account for DockerHub. Not sure if there's something for OWASP and whether we can create a shared secret here.

**Comment by duncanhealy** [comment]
> @duncanhealy makes sense to update the PR then to push the docker image to GitHub's Registry?
> 
> Not sure why did we add this:
> ` if: ${{ secrets.DOCKERHUB_USERNAME=="owasp" }}`
> I'd probably need to use my own private account for DockerHub. Not sure if there's something for OWASP and whether we can create a shared secret here.

It is a good idea to put in conditional checks like that in the action so that the steps don't run on forks of the project
```
if: github.event.pull_request.head.repo.full_name == github.repository
```

before all steps would work

can change above to whichever dockerhub user you choose (owasp do have the dockerhub owasp user)

Was waiting for you to set up the secret there - I put the step if to conditionally run the dockerhub push (it was failing there) 

the push to githubs package registry should run with the existing secrets,

**Comment by lirantal** [comment]
Let me check about using an official owasp/ repository on github

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      -
+        name: Set up QEMU
+        uses: docker/setup-qemu-action@v1
+      -
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Login to DockerHub
+        uses: docker/login-action@v1
+        if: ${{ secrets.DOCKERHUB_USERNAME=="owasp" }}
```
Comment (lirantal) [inline .github/workflows/deploy.yml:64]:
```suggestion
        if: ${{ secrets.DOCKERHUB_USERNAME=="lirantal" }}
```

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+      -
+        name: Build and push
+        uses: docker/build-push-action@v2
+        if: ${{ secrets.DOCKERHUB_USERNAME=="owasp" }}
```
Comment (lirantal) [inline .github/workflows/deploy.yml:71]:
```suggestion
        if: ${{ secrets.DOCKERHUB_USERNAME=="lirantal" }}
```

File: README.md [hunk 3]
```diff
@@ -29,16 +28,37 @@ npx cwe-tool [...command-line options...]
 
 ## Docker
 
-Build and run in one step
+
+### Pull the image from Docker Hub
+
+```shell
+docker pull owasp/cwe-tool
```
Comment (lirantal) [inline README.md:35]:
```suggestion
docker pull lirantal/cwe-tool
```

File: README.md [hunk 4]
```diff
@@ -29,16 +28,37 @@ npx cwe-tool [...command-line options...]
...
+docker run --rm owasp/cwe-tool --search test
```
Comment (lirantal) [inline README.md:36]:
```suggestion
docker run --rm lirantal/cwe-tool --search test
```

**Review by lirantal — CHANGES_REQUESTED** [review]
Updating the docker hub account to my own since we don't yet have anything from OWASP

**Comment by lirantal** [comment]
Let's merge and see how it goes.

**Comment by lirantal** [comment]
@duncanhealy I had to revert this as I couldn't get the github action to get processed even (see if you can examine the github actions build logs). Happy to receive an updated one that works.

**Pull request merged**